### ISSUE TITLE
Template instantiation fails to populate object methods

### ISSUE THREAD

**Issue by llvm-beanz** [microsoft/DirectXShaderCompiler#4315]
DXC lazy-initializes the member methods for built-in objects. This triggers during parsing when objects of an HLSL type are provided, and fails to trigger during template instantiation as shown in the example below. This test case was reduced from the test provided in #4263, but is a separate issue from the "partial specialization" error.

```cpp
template <typename T> struct MyTex2D {
  uint heapId;

  template <typename Arg0> T Load(Arg0 arg0) { return Get().Load(arg0); }

  Texture2D<T> Get() { return (Texture2D<T>)ResourceDescriptorHeap[heapId]; }
};

cbuffer constantBuffer : register(b0) {
  MyTex2D<float4> tex;
};

float4 main() : SV_Target {
  float4 output = tex.Load(float3(0, 0, 0));
  return output;
}
```

**Comment by llvm-beanz** [comment]
The problem is basically that template instantiations are performed as-needed and lazily. This means that `MyTex2D<float4>::Load` is instantiated before `MyTex2D<float4>::Get`. Since `Get` is the function that first creates the `Texture2D<float4>` declaration, the lazy initialization of `Texture2D<float4>`'s members doesn't occur before `Load` is instantiated. This results in `Texture2D<float4>Load` failing to resolve.

This can be resolved by not lazy-initializing the `Texture2D` members, but my initial testing indicates that might be a significant performance degradation.

I'll continue investigating alternative solutions.
